### ISSUE TITLE
StopOnError does not work for parallell tests

### ISSUE THREAD

**Issue by OsirisTerje** [nunit/nunit#3466]
NUnit engine version 3.11

For more information:
See https://github.com/nunit/nunit3-vs-adapter/issues/675#issuecomment-582573234

**Comment by CharliePoole** [comment]
Somebody should complete this story...

"When StopOnError is passed in settings...

"And tests are running in parallel...

"And a single test fails...

"[Say what should happen]"

I think there are alternatives and a new contributor may pick this up.

WRT the console, I think it would be a behavior change if the setting affected multiple assemblies when just one fails.

**Comment by OsirisTerje** [comment]
FWIW:  In the adapter, I'm treating The StopOnError to apply for all assemblies.  It will do a Cancel on the test run after the first failed test responds.  Seems to me that is the logical thing to do.

**Comment by CharliePoole** [comment]
I could be wrong, but I don't believe the console has ever done that.

WRT the adapter, doesn't that implementation only work if MS uses the same process for all assemblies? I assume they don't do that for mixed framework and core targeting.

**Comment by Overclock303** [comment]
Hi, the StopOnError works nice for my use case in sequence mode only. But in parallel i can't use it.
Actually, i don't have the capability to dispose or managed my external resources before the engine cancel all the fixtures.
It does not happen in sequence mode because my custom cleaner do his job on the current test before the engine stop.
I think the fix here is to have [OneTimeBeforeCancel] applied to the fixture and [OneTimeSetUpBeforeCancel] applied to the SetUpFixture.
That way i could manage the way i want my parallel fixtures to run with selenium grid.

Also, i found a bug in parallel mode. When the engine cancel the runs, the tests stops, but the engine is still running.
I need to stop the tests adapter myself. i use the NUnit Adapter 3.17.0.0.

A nice enhancement too, would be the stopOnError to be active only when i decide to raise a fatal exception (a ressource is missing, or ...)
and not only if a logical test Failed. That way i can let the "minor" failure execute the next fixtures, or if a fatal failure occurs i decide to stop all in a managed way.

Also, i think stopOnerror should not applied to all assemblies because, in parallel execution, you can have some assemblies on error, but others just fine.
I play very large business and logical non-regression tests with NUnit and my own framework, and it works like a charm in both mode.
I like the way NUnit is going.

**Comment by CharliePoole** [comment]
Looks like a "mother" issue for NUnit - i.e. a number of separate ideas. :smiley_cat: 

I'm don't work on the framework, so I'll address only your last note...

> _Also, i found a bug in parallel mode. When the engine cancel the runs, the tests stops, but the engine is still running.
I need to stop the tests adapter myself. i use the NUnit Adapter 3.17.0.0._

When running under the adapter, the calling sequence is `vs => adapter '> engine => framework`. That takes place separately __for each assembly__. When a call comes along to cancel the run, the engine cancels the assembly it is running. That assembly should be unloaded as a result. However, a bug is certainly possible, since since the adapter, engine and framework are all running in the same process in this context. After cancelling, the adapter goes back to waiting for the next instruction from VS.

Bottom line: if things are not unloading under the adapter, that's probably an adapter issue, not the NUnit framework or engine. When running under the console, it's different because the engine is then running all test assemblies together.

**Comment by Overclock303** [comment]
i think it s an adapter issue. About  the "mother issue"  yeah, i got this idea while coding last night and block on stopOnError which is a nice feature but can do more. If other guys read me...  thx Charlie ^^ take care

**Comment by CharliePoole** [comment]
IIRC StopOnError is handled as equivalent to a user cancel. I think I did it that way figuring it was the simplest thing that could possibly work (STTCPW) and that we could improve it if and when the need arose. It never arose until we had parallel execution and people started using it. Maybe now is the time to recognize StopOnError as a special case rather than a normal cancel.

**Comment by Overclock303** [comment]
it s definitly a special case with the parallel execution because you cannot dispose anything, that make the feat useless in parallel. 
the assemblies range is also a concern.

**Comment by CharliePoole** [comment]
Right. But it's not treated specially. It's treated as a cancel, which it really shouldn't be. Your suggested attributes would seem to treat it as a special kind of cancel, based on the names, but maybe it shouldn't be viewed as a cancel but rather as a specially requested, orderly shutdown. If that were done, you may not need a special attribute since your normal teardowns would run.

In the framework, user-oriented cancellations are very tricky because they can occur at any time... during a test, during a teardown, etc. But this event can only happen on completion of a test, which should make it possible to handle it without cancelling anything.

If I were doing it, I'd set a `Stopping` flag (new) on all worker queues indicating that no new tests may be enqueued and existing queue entries should be flushed __except__ for onetime teardown tasks, which must continue to execute.

As the workers dequeued items, they would execute onetime teardowns but ignore any tests.

I believe the code already keeps track of "in flight" fixtures - those that have started but not completed. If not, it's easy to add. After all queues are empty, onetime teardown tasks may be staged for those fixtures that are still active.

This is straightforward even though it's somewhat involved. It's definitely no worse than the level of complexity already present in the parallel dispatcher. If I'm right, this would allow StopOnFailure to simply work without special coding by the user.

Of course, any of the teardowns could hang, making it necessary for the user to cancel the run. But that's already the case and loss of information when doing a manual cancel is at least understandable.
